File: ucf_pt/local2025QualifyingRoundData/solutions/roads/roads.py
# UCF Local Contest 2025 (Qualifying) - Country Roads
# Solution written by Tyler Marks

# Scan in the number of cities, initial roads,
# and candidate roads
n, m, r = map(int, input().split())

# Initialize the DSU array
dsu = [-1] * n

# Auxilary functions for DSU
def find(x):
    global dsu
    if dsu[x] < 0: return x
    dsu[x] = find(dsu[x])
    return dsu[x]

def join(x, y):
    global dsu
    x = find(x)
    y = find(y)
    if x == y: return False
    if dsu[x] > dsu[y]: x, y = y, x
    dsu[x] += dsu[y]
    dsu[y] = x
    return True

# Scan in inital edges and join those nodes
# in the DSU
for i in range(m):
    u, v = map(int, input().split())
    join(u-1, v-1)

# Scan in the candidate edges
new_edges = []
for i in range(r):
    u, v, w = map(int, input().split())
    new_edges.append((w, u-1, v-1))

# Sort the new_edges to perform kruskals algorithm
new_edges = sorted(new_edges)

# Go through each edge, and if the join is
# successful, add the weight to the answer
res = 0
for w, u, v in new_edges:
    if join(u, v):
        res += w

# Output the answer
print(res)

File: ucf_pt/local2025QualifyingRoundData/solutions/roads/roads.cpp
// UCF Local Contest 2025 (Qualifying) - Country Roads
// Solution written by Tyler Marks

#include <bits/stdc++.h>
using namespace std;

// Disjoint Set Data Structure
struct DSU {
   vector<int> e;
   DSU(int n): e(n, -1) {}
   int find(int x) { return e[x] < 0 ? x : e[x] = find(e[x]); }
   int join(int x, int y) {
      x = find(x), y = find(y);
      if(x == y) return false;
      if(e[x] > e[y]) swap(x, y);
      e[x] += e[y]; e[y] = x;
      return true;
   }
};

int main() {
    // Scan in number of cities, inital edges
    // add candidate edges
    int n, m, r; cin >> n >> m >> r;
    
    // Initialize the DSU
    DSU dsu(n);

    // Scan in the inital edges and join the
    // corresponding nodes together
    for(int i = 0; i < m; i++) {
        int u, v; cin >> u >> v;
        u--, v--;
        dsu.join(u, v);
    }
    
    // Scan in the candidate edges
    vector<array<int, 3>> new_edges(r);
    for(auto &[w, u, v]: new_edges) {
        cin >> u >> v >> w;
        u--, v--;
    }

    // Sort the edges by weight, allowing us
    // to greedily build the remaining edges
    // (essentially using kruskal's algorithm)
    sort(begin(new_edges), end(new_edges));

    // Go through the edges and if joining the
    // incident nodes is successful, add the
    // edge weight to the answer
    int res = 0;
    for(auto [w, u, v]: new_edges)
        if(dsu.join(u, v))
            res += w;

    // Output the answer
    cout << res << '\n';
}